File: scripts/validate_marketplace.py
#!/usr/bin/env python3
"""Validate this repo as a Claude Code plugin marketplace.

Checks:
  - .claude-plugin/marketplace.json exists, is valid JSON, has required fields
  - every plugin entry resolves to a real directory with a valid plugin.json
  - plugin names are unique, kebab-case, and match their manifest
  - every skill has a SKILL.md with `name` and `description` frontmatter
  - skill directory names match the skill's declared name
  - referenced component dirs (commands/agents/hooks) contain the right files

Exit code 0 = clean, 1 = one or more errors.
"""

from __future__ import annotations

import json
import re
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
NAME_RE = re.compile(r"^[a-z0-9]+(?:-[a-z0-9]+)*$")
SEMVER_RE = re.compile(r"^\d+\.\d+\.\d+(?:[-+].*)?$")

errors: list[str] = []
warnings: list[str] = []


def error(msg: str) -> None:
    errors.append(msg)


def warn(msg: str) -> None:
    warnings.append(msg)


def load_json(path: Path) -> dict | None:
    try:
        with path.open(encoding="utf-8") as fh:
            data = json.load(fh)
    except FileNotFoundError:
        error(f"{rel(path)}: file not found")
        return None
    except json.JSONDecodeError as exc:
        error(f"{rel(path)}: invalid JSON ({exc})")
        return None
    if not isinstance(data, dict):
        error(f"{rel(path)}: top level must be a JSON object")
        return None
    return data


def rel(path: Path) -> str:
    try:
        return str(path.relative_to(ROOT))
    except ValueError:
        return str(path)


def parse_frontmatter(path: Path) -> dict[str, str] | None:
    """Parse the leading `---` YAML block. Only scalar `key: value` pairs are needed."""
    text = path.read_text(encoding="utf-8")
    if not text.startswith("---"):
        error(f"{rel(path)}: missing YAML frontmatter (file must start with '---')")
        return None
    parts = text.split("---", 2)
    if len(parts) < 3:
        error(f"{rel(path)}: frontmatter block is not closed with '---'")
        return None

    fields: dict[str, str] = {}
    key: str | None = None
    for line in parts[1].splitlines():
        if not line.strip() or line.lstrip().startswith("#"):
            continue
        if line[:1].isspace() and key:  # folded continuation
            fields[key] += " " + line.strip()
            continue
        if ":" not in line:
            error(f"{rel(path)}: unparseable frontmatter line: {line!r}")
            return None
        key, value = line.split(":", 1)
        key = key.strip()
        fields[key] = value.strip().strip("'\"")
    return fields


def validate_skill(skill_dir: Path, plugin_name: str) -> None:
    skill_md = skill_dir / "SKILL.md"
    if not skill_md.is_file():
        error(f"{rel(skill_dir)}: skill directory has no SKILL.md")
        return

    fields = parse_frontmatter(skill_md)
    if fields is None:
        return

    name = fields.get("name")
    description = fields.get("description")

    if not name:
        error(f"{rel(skill_md)}: frontmatter is missing required field 'name'")
    else:
        if not NAME_RE.match(name):
            error(f"{rel(skill_md)}: skill name {name!r} must be kebab-case")
        if name != skill_dir.name:
            error(
                f"{rel(skill_md)}: skill name {name!r} does not match "
                f"directory name {skill_dir.name!r}"
            )

    if not description:
        error(f"{rel(skill_md)}: frontmatter is missing required field 'description'")
    elif len(description) > 1024:
        warn(f"{rel(skill_md)}: description is {len(description)} chars; keep it under 1024")

    print(f"    skill: {skill_dir.name} ({plugin_name})")


def validate_plugin(entry: dict, index: int, seen: set[str]) -> None:
    label = f"marketplace.json plugins[{index}]"

    name = entry.get("name")
    if not isinstance(name, str) or not name:
        error(f"{label}: missing required field 'name'")
        return
    if not NAME_RE.match(name):
        error(f"{label}: plugin name {name!r} must be kebab-case")
    if name in seen:
        error(f"{label}: duplicate plugin name {name!r}")
    seen.add(name)

    if not entry.get("description"):
        error(f"{label} ({name}): missing required field 'description'")

    source = entry.get("source")
    if not source:
        error(f"{label} ({name}): missing required field 'source'")
        return
    if not isinstance(source, str):
        # object sources (git/github) point outside this repo; nothing local to check
        print(f"  plugin: {name} (external source, skipped)")
        return

    plugin_dir = (ROOT / source).resolve()
    if not plugin_dir.is_dir():
        error(f"{label} ({name}): source {source!r} is not a directory")
        return

    print(f"  plugin: {name} -> {source}")

    manifest_path = plugin_dir / ".claude-plugin" / "plugin.json"
    manifest = load_json(manifest_path)
    if manifest is not None:
        if manifest.get("name") != name:
            error(
                f"{rel(manifest_path)}: name {manifest.get('name')!r} does not match "
                f"marketplace entry {name!r}"
            )
        if not manifest.get("description"):
            error(f"{rel(manifest_path)}: missing required field 'description'")
        version = manifest.get("version")
        if version and not SEMVER_RE.match(str(version)):
            error(f"{rel(manifest_path)}: version {version!r} is not semver")

    version = entry.get("version")
    if version and not SEMVER_RE.match(str(version)):
        error(f"{label} ({name}): version {version!r} is not semver")

    skills_dir = plugin_dir / "skills"
    if skills_dir.is_dir():
        skill_dirs = sorted(d for d in skills_dir.iterdir() if d.is_dir())
        if not skill_dirs:
            warn(f"{rel(skills_dir)}: directory exists but contains no skills")
        for skill_dir in skill_dirs:
            validate_skill(skill_dir, name)
        stray = [f for f in skills_dir.iterdir() if f.is_file() and f.name == "SKILL.md"]
        if stray:
            error(
                f"{rel(skills_dir)}: SKILL.md must live in a named subdirectory, "
                "not directly under skills/"
            )

    commands_dir = plugin_dir / "commands"
    if commands_dir.is_dir() and not list(commands_dir.rglob("*.md")):
        warn(f"{rel(commands_dir)}: directory exists but contains no .md commands")

    agents_dir = plugin_dir / "agents"
    if agents_dir.is_dir():
        for agent_md in sorted(agents_dir.rglob("*.md")):
            fields = parse_frontmatter(agent_md)
            if fields is None:
                continue
            for required in ("name", "description"):
                if not fields.get(required):
                    error(f"{rel(agent_md)}: frontmatter is missing '{required}'")

    hooks_path = plugin_dir / "hooks" / "hooks.json"
    if hooks_path.exists():
        load_json(hooks_path)

    if not any(
        (plugin_dir / d).is_dir() for d in ("skills", "commands", "agents", "hooks")
    ):
        warn(f"{rel(plugin_dir)}: plugin has no skills, commands, agents, or hooks")


def main() -> int:
    marketplace_path = ROOT / ".claude-plugin" / "marketplace.json"
    print(f"Validating marketplace: {rel(marketplace_path)}")

    marketplace = load_json(marketplace_path)
    if marketplace is None:
        report()
        return 1

    name = marketplace.get("name")
    if not isinstance(name, str) or not name:
        error("marketplace.json: missing required field 'name'")
    elif not NAME_RE.match(name):
        error(f"marketplace.json: name {name!r} must be kebab-case")

    owner = marketplace.get("owner")
    if not isinstance(owner, dict) or not owner.get("name"):
        error("marketplace.json: 'owner' must be an object with a 'name'")

    plugins = marketplace.get("plugins")
    if not isinstance(plugins, list) or not plugins:
        error("marketplace.json: 'plugins' must be a non-empty array")
        report()
        return 1

    seen: set[str] = set()
    for index, entry in enumerate(plugins):
        if not isinstance(entry, dict):
            error(f"marketplace.json plugins[{index}]: entry must be an object")
            continue
        validate_plugin(entry, index, seen)

    report()
    return 1 if errors else 0


def report() -> None:
    print()
    for w in warnings:
        print(f"warning: {w}")
    for e in errors:
        print(f"error: {e}")
    print()
    if errors:
        print(f"FAILED — {len(errors)} error(s), {len(warnings)} warning(s)")
    else:
        print(f"OK — 0 errors, {len(warnings)} warning(s)")


if __name__ == "__main__":
    sys.exit(main())

File: scripts/validate-marketplace.ts
#!/usr/bin/env -S npx tsx
/**
 * Validate this repo as a Claude Code plugin marketplace.
 *
 * Checks:
 *   - .claude-plugin/marketplace.json exists, is valid JSON, has required fields
 *   - every plugin entry resolves to a real directory with a valid plugin.json
 *   - plugin names are unique, kebab-case, and match their manifest
 *   - every skill has a SKILL.md with `name` and `description` frontmatter
 *   - skill directory names match the skill's declared name
 *   - referenced component dirs (commands/agents/hooks) contain the right files
 *
 * Exit code 0 = clean, 1 = one or more errors.
 */

import { readFileSync, readdirSync, statSync } from "node:fs";
import { dirname, join, relative, resolve } from "node:path";
import { fileURLToPath } from "node:url";

const ROOT = resolve(dirname(fileURLToPath(import.meta.url)), "..");
const NAME_RE = /^[a-z0-9]+(?:-[a-z0-9]+)*$/;
const SEMVER_RE = /^\d+\.\d+\.\d+(?:[-+].*)?$/;

const errors: string[] = [];
const warnings: string[] = [];

const error = (msg: string): void => void errors.push(msg);
const warn = (msg: string): void => void warnings.push(msg);
const rel = (path: string): string => relative(ROOT, path) || ".";

/** A `plugins[]` entry in marketplace.json. Every field is unknown until checked. */
interface PluginEntry {
  name?: unknown;
  source?: unknown;
  description?: unknown;
  version?: unknown;
}

interface Marketplace {
  name?: unknown;
  owner?: unknown;
  plugins?: unknown;
}

interface PluginManifest {
  name?: unknown;
  description?: unknown;
  version?: unknown;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function isDir(path: string): boolean {
  try {
    return statSync(path).isDirectory();
  } catch {
    return false;
  }
}

function isFile(path: string): boolean {
  try {
    return statSync(path).isFile();
  } catch {
    return false;
  }
}

function loadJson(path: string): Record<string, unknown> | null {
  let text: string;
  try {
    text = readFileSync(path, "utf8");
  } catch {
    error(`${rel(path)}: file not found`);
    return null;
  }

  let data: unknown;
  try {
    data = JSON.parse(text);
  } catch (exc) {
    error(`${rel(path)}: invalid JSON (${(exc as Error).message})`);
    return null;
  }

  if (!isRecord(data)) {
    error(`${rel(path)}: top level must be a JSON object`);
    return null;
  }
  return data;
}

/** Parse the leading `---` YAML block. Only scalar `key: value` pairs are needed. */
function parseFrontmatter(path: string): Record<string, string> | null {
  const text = readFileSync(path, "utf8");
  if (!text.startsWith("---")) {
    error(`${rel(path)}: missing YAML frontmatter (file must start with '---')`);
    return null;
  }

  const end = text.indexOf("\n---", 3);
  if (end === -1) {
    error(`${rel(path)}: frontmatter block is not closed with '---'`);
    return null;
  }

  const fields: Record<string, string> = {};
  let key: string | null = null;

  for (const line of text.slice(3, end).split("\n")) {
    if (line.trim() === "" || line.trimStart().startsWith("#")) continue;

    if (/^\s/.test(line) && key !== null) {
      // folded continuation of the previous value
      fields[key] = `${fields[key] ?? ""} ${line.trim()}`;
      continue;
    }

    const colon = line.indexOf(":");
    if (colon === -1) {
      error(`${rel(path)}: unparseable frontmatter line: ${JSON.stringify(line)}`);
      return null;
    }

    key = line.slice(0, colon).trim();
    fields[key] = line
      .slice(colon + 1)
      .trim()
      .replace(/^['"]|['"]$/g, "");
  }

  return fields;
}

function listDirs(path: string): string[] {
  return readdirSync(path)
    .map((entry) => join(path, entry))
    .filter(isDir)
    .sort();
}

function walkFiles(path: string, ext: string): string[] {
  const out: string[] = [];
  for (const entry of readdirSync(path)) {
    const full = join(path, entry);
    if (isDir(full)) out.push(...walkFiles(full, ext));
    else if (entry.endsWith(ext)) out.push(full);
  }
  return out.sort();
}

function validateSkill(skillDir: string, pluginName: string): void {
  const skillMd = join(skillDir, "SKILL.md");
  if (!isFile(skillMd)) {
    error(`${rel(skillDir)}: skill directory has no SKILL.md`);
    return;
  }

  const fields = parseFrontmatter(skillMd);
  if (fields === null) return;

  const dirName = skillDir.split("/").pop() ?? "";
  const { name, description } = fields;

  if (!name) {
    error(`${rel(skillMd)}: frontmatter is missing required field 'name'`);
  } else {
    if (!NAME_RE.test(name)) {
      error(`${rel(skillMd)}: skill name '${name}' must be kebab-case`);
    }
    if (name !== dirName) {
      error(
        `${rel(skillMd)}: skill name '${name}' does not match directory name '${dirName}'`,
      );
    }
  }

  if (!description) {
    error(`${rel(skillMd)}: frontmatter is missing required field 'description'`);
  } else if (description.length > 1024) {
    warn(`${rel(skillMd)}: description is ${description.length} chars; keep it under 1024`);
  }

  console.log(`    skill: ${dirName} (${pluginName})`);
}

function validatePlugin(entry: PluginEntry, index: number, seen: Set<string>): void {
  const label = `marketplace.json plugins[${index}]`;

  const name = entry.name;
  if (typeof name !== "string" || name === "") {
    error(`${label}: missing required field 'name'`);
    return;
  }
  if (!NAME_RE.test(name)) {
    error(`${label}: plugin name '${name}' must be kebab-case`);
  }
  if (seen.has(name)) {
    error(`${label}: duplicate plugin name '${name}'`);
  }
  seen.add(name);

  if (!entry.description) {
    error(`${label} (${name}): missing required field 'description'`);
  }

  const source = entry.source;
  if (source === undefined || source === null || source === "") {
    error(`${label} (${name}): missing required field 'source'`);
    return;
  }
  if (typeof source !== "string") {
    // object sources (git/github) point outside this repo; nothing local to check
    console.log(`  plugin: ${name} (external source, skipped)`);
    return;
  }

  const pluginDir = resolve(ROOT, source);
  if (!isDir(pluginDir)) {
    error(`${label} (${name}): source '${source}' is not a directory`);
    return;
  }

  console.log(`  plugin: ${name} -> ${source}`);

  const manifestPath = join(pluginDir, ".claude-plugin", "plugin.json");
  const manifest = loadJson(manifestPath) as PluginManifest | null;
  if (manifest !== null) {
    if (manifest.name !== name) {
      error(
        `${rel(manifestPath)}: name '${String(manifest.name)}' does not match ` +
          `marketplace entry '${name}'`,
      );
    }
    if (!manifest.description) {
      error(`${rel(manifestPath)}: missing required field 'description'`);
    }
    if (manifest.version !== undefined && !SEMVER_RE.test(String(manifest.version))) {
      error(`${rel(manifestPath)}: version '${String(manifest.version)}' is not semver`);
    }
  }

  if (entry.version !== undefined && !SEMVER_RE.test(String(entry.version))) {
    error(`${label} (${name}): version '${String(entry.version)}' is not semver`);
  }

  // an install resolves the marketplace entry, so a disagreement ships a stale version
  if (
    manifest !== null &&
    entry.version !== undefined &&
    manifest.version !== undefined &&
    String(entry.version) !== String(manifest.version)
  ) {
    error(
      `${label} (${name}): version '${String(entry.version)}' does not match ` +
        `${rel(manifestPath)} version '${String(manifest.version)}'`,
    );
  }

  const skillsDir = join(pluginDir, "skills");
  if (isDir(skillsDir)) {
    const skillDirs = listDirs(skillsDir);
    if (skillDirs.length === 0) {
      warn(`${rel(skillsDir)}: directory exists but contains no skills`);
    }
    for (const skillDir of skillDirs) validateSkill(skillDir, name);

    if (isFile(join(skillsDir, "SKILL.md"))) {
      error(
        `${rel(skillsDir)}: SKILL.md must live in a named subdirectory, ` +
          "not directly under skills/",
      );
    }
  }

  const commandsDir = join(pluginDir, "commands");
  if (isDir(commandsDir) && walkFiles(commandsDir, ".md").length === 0) {
    warn(`${rel(commandsDir)}: directory exists but contains no .md commands`);
  }

  const agentsDir = join(pluginDir, "agents");
  if (isDir(agentsDir)) {
    for (const agentMd of walkFiles(agentsDir, ".md")) {
      const fields = parseFrontmatter(agentMd);
      if (fields === null) continue;
      for (const required of ["name", "description"] as const) {
        if (!fields[required]) {
          error(`${rel(agentMd)}: frontmatter is missing '${required}'`);
        }
      }
    }
  }

  const hooksPath = join(pluginDir, "hooks", "hooks.json");
  if (isFile(hooksPath)) loadJson(hooksPath);

  const componentDirs = ["skills", "commands", "agents", "hooks"];
  if (!componentDirs.some((d) => isDir(join(pluginDir, d)))) {
    warn(`${rel(pluginDir)}: plugin has no skills, commands, agents, or hooks`);
  }
}

function report(): void {
  console.log();
  for (const w of warnings) console.log(`warning: ${w}`);
  for (const e of errors) console.log(`error: ${e}`);
  console.log();
  if (errors.length > 0) {
    console.log(`FAILED — ${errors.length} error(s), ${warnings.length} warning(s)`);
  } else {
    console.log(`OK — 0 errors, ${warnings.length} warning(s)`);
  }
}

function main(): number {
  const marketplacePath = join(ROOT, ".claude-plugin", "marketplace.json");
  console.log(`Validating marketplace: ${rel(marketplacePath)}`);

  const marketplace = loadJson(marketplacePath) as Marketplace | null;
  if (marketplace === null) {
    report();
    return 1;
  }

  const name = marketplace.name;
  if (typeof name !== "string" || name === "") {
    error("marketplace.json: missing required field 'name'");
  } else if (!NAME_RE.test(name)) {
    error(`marketplace.json: name '${name}' must be kebab-case`);
  }

  const owner = marketplace.owner;
  if (!isRecord(owner) || !owner["name"]) {
    error("marketplace.json: 'owner' must be an object with a 'name'");
  }

  const plugins = marketplace.plugins;
  if (!Array.isArray(plugins) || plugins.length === 0) {
    error("marketplace.json: 'plugins' must be a non-empty array");
    report();
    return 1;
  }

  const seen = new Set<string>();
  plugins.forEach((entry, index) => {
    if (!isRecord(entry)) {
      error(`marketplace.json plugins[${index}]: entry must be an object`);
      return;
    }
    validatePlugin(entry as PluginEntry, index, seen);
  });

  report();
  return errors.length > 0 ? 1 : 0;
}

process.exit(main());
